distinguish between FIN and RST on datapath

kkourt opened the pull request:
Consider a pod trying to connect to another pod (valid IP) in a port that no app listens to.

```    
client -> server: SYN
server->client: ACK,RST
```
    
This exchange will result in the following CT entry:
    
```
TCP OUT 10.11.0.34:55510 -> 10.11.1.136:6767 expires=335419 (in 21591 sec(s) from now) RxPackets=1 RxBytes=54 RxFlagsSeen=0x04 LastRxReport=313819 TxPackets=1 TxBytes=74 TxFlagsSeen=0x02 LastTxReport=313819 Flags=0x0011 [ RxClosing SeenNonSyn ] RevNAT=0 SourceSecurityID=12121 IfIndex=0
```
   
That is, even though we get an RST flag on the return path, the CT verdict is a reply and the entry is updated with a lifetime of 21600s. This PR distinguishes between FIN and RST TCP and in the case of an RST on the return path it sets the timeout to CT_CLOSE_TIMEOUT (10s).

There are also some additional helper patches so please review them individually.

Update: use BPF_PROG_TEST_RUN tests.

Fixes: #11742

kkourt committed:
datapath/ct: terminate connections that were reset

Consider a pod trying to connect to another pod (valid IP) in a port
that no app listens to.

client -> server: SYN
```
CPU 00: MARK 0xfe6867bb FROM 2287 DEBUG: Conntrack lookup 1/2: src=10.11.0.34:55510 dst=10.11.1.136:6767
CPU 00: MARK 0xfe6867bb FROM 2287 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=1
CPU 00: MARK 0xfe6867bb FROM 2287 DEBUG: CT verdict: New, revnat=0
CPU 00: MARK 0xfe6867bb FROM 2287 DEBUG: Successfully mapped addr=10.11.1.136 to identity=12121
CPU 00: MARK 0xfe6867bb FROM 2287 DEBUG: Conntrack create: proxy-port=0 revnat=0 src-identity=12121 lb=0.0.0.0
```

server->client: ACK,RST
```
CPU 00: MARK 0x0 FROM 2287 DEBUG: No message, arg1=7272 (0x1c68) arg2=1 (0x1)
CPU 00: MARK 0x0 FROM 2287 DEBUG: Conntrack lookup 1/2: src=10.11.1.136:6767 dst=10.11.0.34:55510
CPU 00: MARK 0x0 FROM 2287 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=0
CPU 00: MARK 0x0 FROM 2287 DEBUG: CT entry found lifetime=313879, revnat=0
CPU 00: MARK 0x0 FROM 2287 DEBUG: CT verdict: Reply, revnat=0
```

This exchange will result in the following CT entry:

```
TCP OUT 10.11.0.34:55510 -> 10.11.1.136:6767 expires=335419 (in 21591
sec(s) from now) RxPackets=1 RxBytes=54 RxFlagsSeen=0x04
LastRxReport=313819 TxPackets=1 TxBytes=74 TxFlagsSeen=0x02
LastTxReport=313819 Flags=0x0011 [ RxClosing SeenNonSyn ] RevNAT=0
SourceSecurityID=12121 IfIndex=0
```

That is, even though we get an RST flag on the return path, the CT
verdict is a reply and the entry is updated with a lifetime of 21600s

This patch distinguishes between FIN and RST TCP flags and in the case
of an RST on the return path it sets the timeout to CT_CLOSE_TIMEOUT
(10s).

A test in `bpf-prog-test-run` checks that this is indeed what happens.

One thing to note is that we only set the CT_CLOSE_TIMEOUT only if the
observed TCP flags do not include both SYNs, i.e., only for the cases
where the 3-way handshake is still in progress.

Signed-off-by: Kornilios Kourtis <kornilios@isovalent.com>

kkourt committed:
misc datapath CT comment

Signed-off-by: Kornilios Kourtis <kornilios@isovalent.com>

joestringer commented on bpf/lib/conntrack.h:
```diff
@@ -254,14 +255,15 @@ static __always_inline __u8 __ct_lookup(const void *map, struct __ctx_buff *ctx,
 
 			*monitor = TRACE_PAYLOAD_LEN;
-			if (ct_entry_alive(entry))
+			if (ct_entry_alive(entry) && action != ACTION_RESET)
```
I briefly played with the idea of whether this would be clearer with something like the following instead:

```
static __always_inline void ct_entry_terminate(const struct ct_entry *entry)
{
        entry->rx_closing = entry->tx_closing = 1;
}
```

then above,

```
case ACTION_RESET:
    ct_entry_terminate(entry);
    /* fallthrough */
case ACTION_CLOSE:
```

Then, no change to this line.

Possible implications:
1. This would actually update the fields in the entry that imply closing, so if any other subsequent logic wanted to switch logic based off that (eBPF or userspace by dumping the entry), then connections being reset would be considered also closed. 
1. Could affect code generation

I think point (1) is a bit of a wash, I don't see any such logic so other than a mild preference for readability, I don't think it's much of a motivation to change it. Point (2) is more interesting, I could see some potential for complexity reduction since this branch would have one less condition to check, but I haven't sat down & tried out the two approaches to see if there's a real difference. It could also turn out that the compiler optimizes any differences away.

So I guess in summary, if you're interested in digging a little deeper, this might be a path to learn a bit more but it's a bit orthogonal to the goal you have for this PR.

kkourt replied:
I thought about this as well.

The reason I opted not to do so is that the sender end has not seen the `RST` yet so it is technically not closing yet. I think this is similar to 1. in that it does not currently seem to matter but that might change in the future.

IMHO, we should optimize for readability rather than complexity in this case. Not because complexity does not matter, but because I think the difference is very small and can change with the surrounding code and the compiler version. That is, if complexity is lower now there is a high chance that a couple of code additions and/or compiler versions this might change.

That being said, I don't mind trying this out.

joestringer commented on bpf/lib/conntrack.h:
```diff
@@ -254,14 +255,15 @@ static __always_inline __u8 __ct_lookup(const void *map, struct __ctx_buff *ctx,
 			}
 			break;
 		case ACTION_CLOSE:
+		case ACTION_RESET:
```
nit: Please add `/* fallthrough */` to document that the cases here are intentional:

```
case ACTION_CLOSE:
    /* fallthrough */
case ACTION_RESET:
```

joestringer commented on bpf/lib/conntrack.h:
```diff
@@ -406,10 +408,13 @@ static __always_inline int ct_lookup6(const void *map,
-			else
+			} else {
 				action = ACTION_CREATE;
+			}
```
As the script pointed out, we use kernel style in the BPF datapath code, so no braces unless necessary. For consistency, please switch over to the approach suggested byt he checkpatch script.

kkourt replied:
Addressed.

joestringer commented on test/k8sT/CT.go:
```diff
@@ -0,0 +1,84 @@
+		kubectl = helpers.CreateKubectl(helpers.K8s1VMName(), logger)
+		ciliumFilename = helpers.TimestampFilename("cilium.yaml")
+		DeployCiliumOptionsAndDNS(kubectl, ciliumFilename, map[string]string{
+			"bpf.monitorAggregation": "none",
```
Do you need `monitorAggregation: none` to validate this functionality? Surely the "new" "RST" flag should always trigger a monitor notification, even if monitor aggregation is set to medium (default) ?

kkourt replied:
nope, it's a leftover. I'll remove it.

joestringer commented on test/k8sT/CT.go:
```diff
@@ -0,0 +1,84 @@
+			ciliumPodK8s1, err := kubectl.GetCiliumPodOnNode(helpers.K8s1)
+			ExpectWithOffset(2, err).Should(BeNil(), "Cannot get cilium pod on k8s1")
+			// ciliumPodK8s2, err := kubectl.GetCiliumPodOnNode(helpers.K8s2)
+			// ExpectWithOffset(2, err).Should(BeNil(), "Cannot get cilium pod on k8s2")
```
Leftover code to be removed?

joestringer commented on test/k8sT/manifests/testpods.yaml:
```diff
@@ -0,0 +1,36 @@
+---
```
Is there something unique / different about these pods vs. some of the other sample apps we have? I would prefer to minimize the number of new different manifests we use if possible, so that if we need to debug the test, we don't have to spend as much time understanding what the manifest does compared with the deployments we use in other tests. Of course, feel free to disregard this point if there's some obvious difference in this manifest that I'm missing.

kkourt replied:
I could not find any, but maybe I missed it? The distinguish characteristic is that these are pods (i.e., not deployments) and they use the netperf container (technically, netcat but we typically use the netperf container for that). 

The closest one I found was `test-cli.yaml` but this one has the pods running on the same node instead of different nodes which was what I wanted. (it probably does not make a difference, but since we are getting two nodes up anyway, I felt it's safer to test with packets going over the wire.)

joestringer commented on test/k8sT/CT.go:
```diff
@@ -0,0 +1,84 @@
+	"github.com/cilium/cilium/test/helpers"
+)
+
+var _ = Describe("K8sConnectionTracking", func() {
```
I wonder, can we achieve the same testing through `bpf/tests/conntrack_test.h` `test___ct_lookup()` instead? If we're just trying to validate timeout behaviour, that's one of the few things we actually have unit test infrastructure that we can use to validate.

It should take fractions of a second to validate the change in functionality, not seconds or minutes worth of configuring the cluster, creating apps, running commands to generate traffic, just to check a timeout.

If the goal of this is to put some infrastructure in place to be able to build out more comprehensive testing, or if there's unusual application behaviour that requires integration testing to properly simulate, then I can see arguments to have additional tests like this. Otherwise I wonder if we can gain the same validation in a more lightweight manner.

kkourt replied:
I agree that it is too much overhead for this simple use-case. I also think that the unit tests will not have the same coverage as the current test, since AFAIU we do not process actual packets in the latter.

I think there is a (mostly) happy middle ground where we:  create a tap device, load parts (e.g., the CT code with some minimal initial processing) of the datapath, send constructed packets to the interface, and check the state from the BPF maps. I think that this would generally very useful to have, and I cannot think of a reason why it wouldn't work. I want to try and hack something up and see if I can make it work.

joestringer replied:
There's also I believe the option to load a program, then simulate running a packet through using the `BPF_PROG_TEST_RUN` command to the `bpf()` syscall.

kkourt replied:
Pushed a new version that uses such a test.

joestringer commented on test/k8sT/Services.go:
```diff
@@ -779,7 +779,7 @@ var _ = Describe("K8sServicesTest", func() {
 
 			// Field #7 is "RxPackets=<n>"
-			cmdIn := "cilium bpf ct list global | awk '/%s/ { sub(\".*=\",\"\", $7); print $7 }'"
+			cmdIn := "cilium bpf ct list global | perl -ne 'print $1 if /%s.*RxPackets=(\\d+).*/'"
```
These changes appear to be made in the commit which introduces a new `--time-diff` flag, but these commands aren't using the new flag. Is the default behaviour changing? Should this change be in a different commit?

Also, I would be reluctant to start adding `perl` usage into the codebase simply to minimize the number of different languages we use. I'm happy to be overruled if we think perl knowledge is more common across the Cilium contributor community than awk (they're both a bit foreign to me), but I'll at least raise it as a point here since this seems like this would be the first perl expression.

aanm replied:
+1

kkourt replied:
> These changes appear to be made in the commit which introduces a new --time-diff flag, but these commands aren't using the new flag. Is the default behaviour changing? Should this change be in a different commit?

Makes sense.

> Also, I would be reluctant to start adding perl usage into the codebase simply to minimize the number of different languages we use. I'm happy to be overruled if we think perl knowledge is more common across the Cilium contributor community than awk (they're both a bit foreign to me), but I'll at least raise it as a point here since this seems like this would be the first perl expression.

I thnk the perl one liner above is more easy to understand than the awk above because it is just a single regular expression. Even if you don't understand all the details, I think you get an idea of what it does. That's my personal opinion colored by my personal bias though :-)

kkourt replied:
Change removed.

joestringer reviewed (changes requested):
I didn't look closely at the --time-diff command but it looks super useful. I'll let other reviewers look closer at that area.

A few minor points on the datapath and testing side. In particular, it seems like integration testing is a heavy hammer to use for validating functionality that should be able to be unit-tested. More detail below.

aanm commented on cilium/cmd/bpf_ct_list.go:
```diff
@@ -57,6 +63,36 @@ func getMaps(eID string) []*ctmap.Map {
+	var err error
+
+	if timeDiff {
+		params := daemon.NewGetHealthzParamsWithTimeout(5 * time.Second)
```
Usually the `cilium bpf` commands don't require connection to the agent. Can't the developer pass a flag with the clock type as a fallback?

kkourt replied:
I thought about this option, but I decided that it would not be very ergonomic to do so and opted to make their (and our) lives easier. Especially because users would have to use the same first-level command (`cilium`) to do this:
```
cilium status --verbose
cilium bpf ct list --time-diff --clock-source xxx
```

I believe connecting the agent _only_ if a specific flag is give is a good tradeoff because the tool will continue to work on its basic form (as before) without depending on the agent.

aanm replied:
> I thought about this option, but I decided that it would not be very ergonomic to do so and opted to make their (and our) lives easier. Especially because users would have to use the same first-level command (`cilium`) to do this:
> 
> ```
> cilium status --verbose
> cilium bpf ct list --time-diff --clock-source xxx
> ```
> 
> I believe connecting the agent _only_ if a specific flag is give is a good tradeoff because the tool will continue to work on its basic form (as before) without depending on the agent.

What I'm suggesting is to have a flag that allows someone to pass the `time-diff` as a value of its own, this is useful even if the agent is not available. I'm not saying to replace the `params := daemon.NewGetHealthzParamsWithTimeout(5 * time.Second)` with the flag.

kkourt replied:
I see, so

```
cilium bpf ct list --time-diff
```

contacts the server, and:


```
cilium bpf ct list --time-diff --time-diff-clocksource=jiffies|ktime
```

does not contact the server.

Makes sense!

kkourt replied:
Updated version includes this change.

aanm commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -283,6 +284,54 @@ func (m *Map) DumpEntries() (string, error) {
+			return fmt.Sprintf("remaining: %d sec(s)", diff)
+		}
+
+	case "jiffies":
```
`case models.ClockSourceModeJiffies:`

aanm commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -283,6 +284,54 @@ func (m *Map) DumpEntries() (string, error) {
+func DumpEntriesWithTimeDiff(m CtMap, clockSource models.ClockSource) (string, error) {
+	var toRemSecs func(uint32) string
+	switch clockSource.Mode {
+	case "ktime":
```
`case models.ClockSourceModeKtime:`

aanm commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -283,6 +284,54 @@ func (m *Map) DumpEntries() (string, error) {
+	}
+	// DumpWithCallback() must be called before sb.String().
+	err := m.DumpWithCallback(cb)
+	return sb.String(), err
```
if there's an error, do we still need to do `sb.String`?, i.e.:
```
err := m.DumpWithCallback(cb)
if err != nil {
    return "", err
}
return sb.String(), nil
```

pchaigno replied:
@kkourt I think you forgot to push the change to address :point_up:

kkourt replied:
Indeed. I changed one of them but not both. Should be fixed now.

christarazi commented on cilium/cmd/bpf_ct_list.go:
```diff
@@ -57,6 +63,36 @@ func getMaps(eID string) []*ctmap.Map {
 
+func doDumpEntries(m ctmap.CtMap) {
+	var out string
+	var err error
```
Nit:

```
var (
 out string
 err  error
)
```

kkourt replied:
fixed

kkourt committed:
bpf: add BPF_PROG_RUN_TEST tests

This patch introduces bpf testing facilities using BPF_PROG_RUN_TEST. We
use the cilium/ebpf library to load and test the programs.

cilium/ebpf does not seem to support  the .pinning field for maps when
they are read from the corresponding ELF section. Not sure how easy it
would be to add such support, but get something working without changing
ebpf we patch the ELF file and zero the unsupported fields out.

There are three tests currently. The first two are just sanity checks
that everything works as expected. The last one prints the timeout of a
CT entry when a SYN TCP packet is sent and a TCP RST packet is received.
It is intended for a subsequent patch.

Use go test to run the tests instead of doing so from main.  The tests
can now be invoked by `make go_prog_test` on the bpf/ directory.

```
$ make go_prog_test
clang -DENABLE_IPV4 -DENABLE_IPV6 -DENABLE_HOST_REDIRECT -DENABLE_NAT46 -DENABLE_ROUTING -DENABLE_IPSEC -DPOLICY_VERDICT_NOTIFY -I/home/kkourt/go/src/github.com/cilium/cilium/bpf/include -I/home/kkourt/go/src/github.com/cilium/cilium/bpf -D__NR_CPUS__=8 -O2 -g -target bpf -std=gnu89 -nostdinc -emit-llvm -Wall -Wextra -Werror -Wshadow -Wno-address-of-packed-member -Wno-unknown-warning-option -Wno-gnu-variable-sized-type-not-at-end -Wdeclaration-after-statement -c tests/bpf_ct_tests.c -o tests/bpf_ct_tests.ll
llc -march=bpf -mcpu=probe -mattr=dwarfris -filetype=obj -o tests/bpf_ct_tests.o tests/bpf_ct_tests.ll
sudo go test ./tests/prog_test
ok      github.com/cilium/cilium/bpf/tests/prog_test    0.048s
```

Signed-off-by: Kornilios Kourtis <kornilios@isovalent.com>

tgraf commented on bpf/lib/conntrack.h:
```diff
@@ -254,14 +255,15 @@ static __always_inline __u8 __ct_lookup(const void *map, struct __ctx_buff *ctx,
 
 			*monitor = TRACE_PAYLOAD_LEN;
-			if (ct_entry_alive(entry))
+			if (ct_entry_alive(entry) && action != ACTION_RESET)
```
Doesn't this allow for anybody to spam a ton of RSTs (with invalid sequence numbers) in one direction without requiring the victim to have a chance to react and cause all lifetimes of the conntrack table to be reset to 10s to trigger the expiration of connections that could still be alive?

kkourt replied:
The packets would have to be mapped to the same CT entry, so AFAIU they would need to have the same 4-tuple.

When we originally discussed this approach on the datapath meeting, I believe that @joestringer mentioned that we do not allow pods to send packets that do not match their IP. So the spurious packets will have to come from the same pod as the alive connections, which under normal operation should not happen since the port should already be binded if an active connection exists.

kkourt replied:
If I'm missing something, one way to amortize the problem would be to only set `ACTION_RESET` for CT entries that have not seen both `SYN`s, since from a quick look we are already tracking this. In this case, the issue would happen only during the 3-way handshake.

tgraf replied:
> When we originally discussed this approach on the datapath meeting, I believe that @joestringer mentioned that we do not allow pods to send packets that do not match their IP. So the spurious packets will have to come from the same pod as the alive connections, which under normal operation should not happen since the port should already be binded if an active connection exists.

Agreed but I don't think that is protection enough to assume that we are fully OK with anybody with access to schedule a host networking pod with raw socket access can terminate connections of other pods at will.

> If I'm missing something, one way to amortize the problem would be to only set ACTION_RESET for CT entries that have not seen both SYNs, since from a quick look we are already tracking this. In this case, the issue would happen only during the 3-way handshake.

Yes, that is a good protection mechanism. The handshake makes it reasonable to expect packets delivered in both directions so even if the lifetime is adjusted incorrectly by a malicious packet, it would get reset when the SYN-ACK arrives. The main concern is that the lifetime can be reduced on connections which are iding.

kkourt replied:
Updated version does what is described above, i.e., only reduces the timeout for connections where not both SYNs have been seen.

kkourt committed:
ctmap: add --time-diff option to show relative time

This patch adds an --time-diff option to show the relative
expiration time (compared to now) when dumping CT entries with `cilium
bpf ct list`.

`cilium bpf ct list --time-diff` contacts the agent to get the current
clock source. Two other options: `--time-diff-clocksource-mode` and
`--time-dif-clocksource-hz` allow users to manually set the clocksource
and avoid contacting the agent.

Output example:

```
ICMP IN 192.168.33.12:0 -> 10.11.0.138:5564 expires=2046 (remaining: -602 sec(s)) RxPackets=1 RxBytes=50 RxFlagsSeen=0x00 LastRxReport=1986 TxPackets=1 TxBytes=50 TxFlagsSeen=0x00 LastTxReport=1986 Flags=0x0000 [ ] RevNAT=0 SourceSecurityID=2 IfIndex=0
ICMP IN 10.11.0.76:0 -> 10.11.0.138:29009 expires=2284 (remaining: -364 sec(s)) RxPackets=1 RxBytes=50 RxFlagsSeen=0x00 LastRxReport=2224 TxPackets=1 TxBytes=50 TxFlagsSeen=0x00 LastTxReport=2224 Flags=0x0000 [ ] RevNAT=0 SourceSecurityID=1 IfIndex=0
ICMP IN 192.168.33.12:0 -> 10.11.0.138:16144 expires=2166 (remaining: -482 sec(s)) RxPackets=1 RxBytes=50 RxFlagsSeen=0x00 LastRxReport=2106 TxPackets=1 TxBytes=50 TxFlagsSeen=0x00 LastTxReport=2106 Flags=0x0000 [ ] RevNAT=0 SourceSecurityID=2 IfIndex=0
ICMP IN 10.11.0.76:0 -> 10.11.0.138:0 related expires=2704 (remaining: 56 sec(s)) RxPackets=1 RxBytes=50 RxFlagsSeen=0x00 LastRxReport=2644 TxPackets=0 TxBytes=0 TxFlagsSeen=0x00 LastTxReport=0 Flags=0x0010 [ SeenNonSyn ] RevNAT=0 SourceSecurityID=1 IfIndex=0
```

Signed-off-by: Kornilios Kourtis <kornilios@isovalent.com>

ti-mo commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,476 @@
+	return ret, nil
+}
+
+func MapDeleteAll(m *ebpf.Map) error {
```
Exported function, but doesn't seem to be used outside this package.

ti-mo commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,476 @@
+package main
```
Why not make this a testable Go package? (e.g. renaming to `bpf/tests/prog_test.go`)

kkourt replied:
That would require two .go files, right? I was just opting for keeping things simple...

joestringer replied:
Nope, not required:

```
$ ls
foo_test.go
$ go test -v ./...
=== RUN   TestIntMinBasic
--- PASS: TestIntMinBasic (0.00s)
=== RUN   TestIntMinTableDriven
=== RUN   TestIntMinTableDriven/0,1
=== RUN   TestIntMinTableDriven/1,0
=== RUN   TestIntMinTableDriven/2,-2
=== RUN   TestIntMinTableDriven/0,-1
=== RUN   TestIntMinTableDriven/-1,0
--- PASS: TestIntMinTableDriven (0.00s)
    --- PASS: TestIntMinTableDriven/0,1 (0.00s)
    --- PASS: TestIntMinTableDriven/1,0 (0.00s)
    --- PASS: TestIntMinTableDriven/2,-2 (0.00s)
    --- PASS: TestIntMinTableDriven/0,-1 (0.00s)
    --- PASS: TestIntMinTableDriven/-1,0 (0.00s)
PASS
ok      _/home/joe/playground/test      0.003s
```

foo_test.go is a direct copy of the code from https://gobyexample.com/testing .

kkourt replied:
Ah OK, now I understand. @ti-mo 's point was to invoke this as a testing package using `go test`. I think that makes sense.

pchaigno replied:
Were there any changes made to address this? What's the `go test` command we should use?

kkourt replied:
Fixed this (see last patch). 

There is now a `make go_prog_test` target that builds the object file and runs the test. It runs `sudo go test ./tests/prog_test`

ti-mo commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,476 @@
+	"golang.org/x/sys/unix"
+)
+
+type Ct4GlobalMap map[ctmap.CtKey4Global]ctmap.CtEntry
```
Exported symbol only used in this package.

ti-mo commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,476 @@
+
+type Ct4GlobalMap map[ctmap.CtKey4Global]ctmap.CtEntry
+
+func Ct4TcpMapFromBPF(ebpfMap *ebpf.Map) (Ct4GlobalMap, error) {
```
Exported function only used in this package.

aanm commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,473 @@
+package main
```
needs copyright header.

kkourt replied:
Thanks!

aanm reviewed (approved):
Minot nit that needs to be addressed but doesn't require my 2nd review.

kkourt committed:
bpf/tests/conntrack_test.h: zero out TCP flags

The previous patch checks for TCP_RST in seen_flags, which makes this
test to fail due to uninitialized data. Zero out the TCP flags before
the test.

Signed-off-by: Kornilios Kourtis <kornilios@isovalent.com>

borkmann commented on bpf/lib/conntrack.h:
```diff
@@ -253,12 +261,20 @@ static __always_inline __u8 __ct_lookup(const void *map, struct __ctx_buff *ctx,
+				entry->rx_closing = 1;
+			} else {
+				entry->tx_closing = 1;
+			}
```
This should only be done for dir != CT_SERVICE since for the latter we may not see both directions.

kkourt replied:
Do you mean that the last `else` should be `else if (dir == CT_EGRESS)` like so:

```
	/* If we got an RST and have not seen both SYNs,
	 * terminate the connection
	 */
	if (!ct_entry_seen_both_syns(entry) &&
	    (seen_flags.value & TCP_FLAG_RST)) {
		entry->rx_closing = 1;
		entry->tx_closing = 1;
	} else if (dir == CT_INGRESS) {
		entry->rx_closing = 1;
	} else if (dir == CT_EGRESS) {
		entry->tx_closing = 1;
	}
```

brb replied:
```
if (!ct_entry_seen_both_syns(entry) && (seen_flags.value & TCP_FLAG_RST) && (dir != CT_SERVICE) {
    entry->rx_closing = 1;
    entry->tx_closing = 1;
} else if ...
```

Otherwise, we would prematurely mark the `CT_SERVICE` entry for removal. In the case of `CT_SERVICE`, we see packets only in one direction (to a service).

kkourt replied:
Got it! Thanks. Should be fixed on the latest version.

tgraf reviewed (approved):
My concern has been addressed so LGTM.

brb commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -286,6 +290,54 @@ func (m *Map) DumpEntries() (string, error) {
 	return DoDumpEntries(m)
 }
 
+func DumpEntriesWithTimeDiff(m CtMap, clockSource models.ClockSource) (string, error) {
```
Have you considered making the `DoDumpEntries(..)` function to accept an option parameter which could contain `bool` to indicate whether time-diff is required and if so, then another field which contains the clock source? That would eliminate a bit of boilerplate code.

kkourt replied:
I've reworked the code to avoid the boilerplate.

brb commented on pkg/maps/ctmap/types.go:
```diff
@@ -588,3 +588,22 @@ func (c *CtEntry) String() string {
 		c.IfIndex)
 }
+
+func (c *CtEntry) TimeDiffString(toRemSecs func(uint32) string) string {
```
Maybe we could change `(c *CtEntry) String(..)` to `StringWithOpt(..)`, then make `String(..)` to invoke `StringWithOpt(..)`, so that we could DRY?

kkourt replied:
Used `StringWithTimeDiff` on latest patch to do so.

pchaigno commented on bpf/Makefile:
```diff
@@ -8,6 +8,7 @@ include ../Makefile.defs
 SUBDIRS = sockops
```
Commit `bpf: add BPF_PROG_RUN_TEST tests` has a typo on `It is intentended for a subsequent patch.` Probably not worth re-spinning just for that.

kkourt replied:
Did a respin anyway, so it should be fixed now.

pchaigno commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,487 @@
+func testMap(spec *ebpf.Collection) error {
+	prog := spec.Programs["test_map"]
+	if prog == nil {
+		return errors.New("did not find test_nop program")
```
```suggestion
		return errors.New("did not find test_map program")
```

kkourt replied:
Fixed! Thanks!

pchaigno commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,487 @@
+		return errors.New("did not find test_nop program")
+	}
+
+	bpfCtMap := spec.Maps["test_cilium_ct_tcp4_65535"]
```
Isn't a nil check needed here or in `ct4TcpMapFromBPF`?

kkourt replied:
Yap! Added the check!

pchaigno commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,487 @@
+			DstMAC:       net.HardwareAddr{1, 0, 0, 3, 0, 20},
+			EthernetType: layers.EthernetTypeIPv4,
+		},
+	)
```
Given the same packet is defined several times, maybe worth declaring it as a variable of the package, not of each test function?

kkourt replied:
~It's not the same packet -- it's source/destination addresses/ports are reversed.~

I don't think it offers much at this point. Once we add more tests, I think we use build packet builder functions, but it's probably out of scope for this PR.

(edited)

pchaigno replied:
> I don't think it offers much

It shortens the code and makes it easier to read. Definitely not a blocker.

pchaigno commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,487 @@
+			"test":  "map",
+			"error": err,
+		}).Fatal("test failed")
+	}
```
I think you can use a function slice for `testNop`, `testMap`, and `testCt4Rst`.

kkourt replied:
Hmm, if I understand you mean I could do something like:

```
for fn := range  testFns {
    fn(...)
}
```

I'd rather not, since we might want to pass different arguments to the test functions eventually.

pchaigno commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,487 @@
+	return nil
+}
+
+func patchElf(fname string, fnameOut string) error {
```
As far as I understand, this patching function is only necessary because we use the map definition convention of iproute2. If we switch to use cilium/ebpf as our loader, it will likely not be necessary anymore (not sure of libbpf's convention if that's what we end up using). Given this patching is kind of a hack, maybe it would be worth clarifying the intent and purpose in a comment?

kkourt replied:
Definitely is, thanks! Added a comment.

> If we switch to use cilium/ebpf as our loader, it will likely not be necessary anymore

I thought that this (i.e., not processing al the fields of `bpf_elf_map`) was something that should be fixed on cilium/ebpf. For the current tests, it does not matter because we  don't care about these fields but for the loader I think we do. Nothing to do with this PR, just thought I'd mention it.

pchaigno commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,487 @@
+		// special record values set from our bpf programs to mark their termination
+		if dm.SubType == monitor.DbgUnspec && dm.Arg1 == 0xe3d && dm.Arg2 == 0xe3d {
+			break
+		}
```
Couldn't you run `dumpDebugMessages` as a goroutine then use a channel to signal the end of `BPF_PROG_TEST_RUN`? it would avoid this hack where we use debug logs to signal the end. We won't be able to use this hack if we later want to extend those `BPF_PROG_TEST_RUN` tests to our existing object files.

kkourt replied:
Yap, Agreed.

I've seen code that already does this somewhere in our tree. I think the main issue is that we do not want to follow this approach when we integrate this test to CI tests, because we don't want to block forever (or wait for the CI timeout).

I would like to add a subsequent PR that adds some more CT tests and integrates this into the CI, so I plan to address it there.

pchaigno commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,487 @@
+			DataOffset: 5,
+			SYN:        true,
+		},
+		gopacket.Payload([]byte{'p', 'i', 'z', 'z', 'a', '!', ' ', ':', '-', ')'}),
```
:smile:

pchaigno commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,487 @@
+			DataOffset: 5,
+			RST:        true,
+		},
+		gopacket.Payload([]byte{'p', 'i', 'z', 'z', 'a', '!', ' ', ':', '-', ')'}),
```
Maybe best to make a copy of the first packet and change the only two fields that differ? It took me a while to spot the difference (before I read the variable's name).

kkourt replied:
It's not the same packet -- its source/destination addresses/ports are reversed.

pchaigno replied:
One more argument to copy the packet IMO, because it's very much not obvious that the ports are reversed currently.

kkourt replied:
Not sure what you mean by copying the packet. Do you mean copy the packet and change the fields by finding their respective offsets? If you could provide an example of what is that you are suggesting, I'd be happy to change the code to match it.

pchaigno replied:
Discussed this offline a bit: there's no good way to modify the packet after creating it with gopacket :disappointed: So it looks like the code duplication is the best we can do for now without big changes.

pchaigno commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,487 @@
+
+	// First Packet: SYN
+	prog1 := spec.Programs["test_ct4_rst1"]
+	if prog1 == nil {
```
I think the Go way of writing that would be, but maybe worth double checking:
```suggestion
	if prog1 := spec.Programs["test_ct4_rst1"]; prog1 == nil {
```
Also works for most other checks in this file.

kkourt replied:
`golint` does not complain about it, but  I've seen it complain it when using this syntax in some cases (although don't remember which and it is entirely possible that I'm misremembering) so I tend not to use it.

pchaigno replied:
Didn't say it was incorrect Go lint :slightly_smiling_face:
I find it makes the code a bit easier to read, especially when its a successions a checks :wink: 
And we're using it a lot in our code base:
```
$ git grep -P "; \w+ != nil" -- "*.go" | grep -v "^vendor\/" | wc -l
2579
```

kkourt replied:
I personally don't find it more readable but it's probably an acquired taste. I'll change it and repush.

pchaigno replied:
It could be worth raising in e.g. our #golang channel to see how folks are usually using it.

kkourt replied:
Pushed a new version with the above syntax. It was not trivial to rebase, so added a new patch on top.

pchaigno commented on bpf/tests/bpf_ct_tests.c:
```diff
@@ -0,0 +1,163 @@
+#include <bpf/ctx/skb.h>
+#include <bpf/api.h>
+
+/* most values taken from node_config.h */
```
Probably naive question: can't we just import `node_config.h`?

kkourt replied:
We could, and that's what I did for an initial version but this way we get more freedom in defining only the definitions we want. 

E.g., if I remember correctly, if we include `node_config.h` we cannot directly define `CONNTRACK`, we need to define either `ENABLE_NODEPORT` or `ENABLE_HOST_FIREWALL`. I prefer just defining `CONNTRACK`.

pchaigno commented on bpf/tests/bpf_ct_tests.c:
```diff
@@ -0,0 +1,163 @@
+	};
+	struct bpf_elf_map *ctmap = &CT_MAP_TCP4;
+
+	(void)ctx;
```
Our convention for unused arguments is to use `__maybe_unused` which is defined in `include/bpf/compiler.h` and looks safe to pull in here.

kkourt replied:
fixed!

brb reviewed (commented):
Thanks, some comments.

This BPF_PROG_TEST_RUN based testing is :rocket:. If you could convert it to _test.go (as suggested by others), and make it a part of CI, it would be :100: :rocket:.

pchaigno commented on bpf/lib/conntrack.h:
```diff
@@ -223,6 +223,7 @@ static __always_inline __u8 __ct_lookup(const void *map, struct __ctx_buff *ctx,
 			ct_state->dsr = entry->dsr;
 			ct_state->proxy_redirect = entry->proxy_redirect;
+			/* See the ct_create4 comments re the rx_bytes hack */
 			if (dir == CT_SERVICE)
```
@brb I just became aware of this hack. Would it be worth writing a small Coccinelle script to check that all assignments of `x != entry->rx_bytes` to `ct_state->backend_id` are guarded by a check that `dir == z; z != CT_SERVICE`? Happy to write it given it's fairly trivial. It would guarantee we have a test failing if the assumption behind this hack ever becomes false.

brb replied:
Yeah, makes sense. Thanks!

pchaigno reviewed (commented):
Hopefully my review comments don't conflict too much with Martynas's :worried: 

The new `BPF_PROG_TEST_RUN` program looks awesome! And it's simpler than I expected :tada: 
Probably worth adding it to the CI or it will become stale very quickly.

I think long term it would be worth to extend this test to be able to apply such checks to our unmodified object files. We would probably also benefit from having a slightly higher-level way of describing input parameters and expected outputs. Maybe best to move that discussion to a dedicated issue.

pchaigno reviewed (commented):
I'm super happy for the introduction of this new `BPF_PROG_TEST_RUN`-based tests, but why couldn't this be tested in `test___ct_lookup`? For this particular use case, it seems a lot simpler...

pchaigno commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,494 @@
+	}
+
+	bpfCtMap := spec.Maps["test_cilium_ct_tcp4_65535"]
+	if prog == nil {
```
```suggestion
	if bpfCtMap == nil {
```

kkourt replied:
:man_facepalming: 

Thanks for catching this :)

kkourt replied:
It should be fixed now.

pchaigno commented on bpf/Makefile:
```diff
@@ -213,6 +213,10 @@ tests/bpf_ct_tests.o: tests/bpf_ct_tests.ll
 
+.PHONY: go_prog_test
+go_prog_test: $(BPF_TEST)
+	sudo go test ./tests/prog_test
```
From the development VM:
```
vagrant@k8s1:~/go/src/github.com/cilium/cilium/bpf$ sudo go test ./tests/prog_test
package github.com/cilium/cilium/bpf/tests/prog_test: build constraints exclude all Go files in /home/vagrant/go/src/github.com/cilium/cilium/bpf/tests/prog_test
```
Same error if I run `make -C bpf go_prog_test` from the root of the Cilium sources.

kkourt replied:
Thanks for testing this! I'll convert the PR to draft until I figure out how to deal with this. (I believe this is due to the build contrain I added). (edited.)

pchaigno commented on bpf/lib/conntrack.h:
```diff
@@ -39,6 +39,14 @@ static __always_inline bool conn_is_dns(__u16 dport)
 	return dport == bpf_htons(53);
```
FYI, the commit `datapath/ct: terminate connections that were reset` triggers a checkpatch warning because it has two signed-off-by.

kkourt replied:
yap, thanks! I've already fixed this into my local tree. I'll push it when I have all fixes ready.

brb reviewed (approved):
Thanks!

christarazi commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -261,25 +262,67 @@ type GCFilter struct {
+// ct entries in m to a string. If clockSource is not nil, it uses it to
+// compute the time difference of each entry from now and prints that too.
+func DumpEntriesWithTimeDiff(m CtMap, clockSource *models.ClockSource) (string, error) {
+
```
Nit: extra line here not needed

christarazi commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,494 @@
+}
+
+func testCt4Rst(spec *ebpf.Collection) error {
+
```
Nit: extra line

kkourt replied:
fixed

christarazi commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,494 @@
+			DataOffset: 5,
+			RST:        true,
+		},
+		gopacket.Payload([]byte{'p', 'i', 'z', 'z', 'a', '!', ' ', ':', '-', ')'}),
```
Not a big deal, but was wondering why not just make a string "pizza! :-)" and then cast to `[]byte`?

kkourt replied:
fixed

christarazi commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,494 @@
+}
+
+func main() {
+
```
Nit: extra lina

kkourt replied:
fixed by squashing

christarazi commented on bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go:
```diff
@@ -0,0 +1,494 @@
+	err := unix.Setrlimit(unix.RLIMIT_MEMLOCK, &lim)
+	if err != nil {
+		log.Fatalf("setrlimit: %v", err)
+	}
```
Why is this needed?

kkourt replied:
It's needed to run the bpf programs.

christarazi commented on bpf/Makefile:
```diff
@@ -8,7 +8,7 @@ include ../Makefile.defs
 SUBDIRS = sockops
 
 BPF_SIMPLE = bpf_network.o bpf_alignchecker.o
```
Typo in commit `invovekd` :)

kkourt replied:
fixed.

christarazi commented on bpf/tests/prog_test/prog_test.go:
```diff
@@ -446,52 +451,52 @@ func testCt4Rst(spec *ebpf.Collection) error {
+	objDir := ".."
+	fname := path.Join(objDir, "bpf_ct_tests.o")
+	fnamePatched := path.Join(objDir, "bpf_ct_tests_patched.o")
+	err := patchElf(fname, fnamePatched)
```
I see that this commit is changing a previous commit. Why not just squash these two commits together?

kkourt replied:
squashed

christarazi commented on bpf/tests/prog_test/prog_test.go:
```diff
@@ -75,8 +75,7 @@ func mapDeleteAll(m *ebpf.Map) error {
 
-		err = m.Delete(key)
-		if err != nil {
+		if err := m.Delete(key); err != nil {
```
Same here re: squashing commits

kkourt replied:
fixed

christarazi reviewed (changes requested):
LGTM overall. Mostly minor nits and some commit squashing comments.
